app/controllers/application_controller.rb

```diff
@@ -3,8 +3,148 @@
 class ApplicationController < Sinatra::Base
 
   configure do
+    enable :sessions
+    set :session_secret, "password_security"
+  end
     set :public_folder, 'public'
     set :views, 'app/views'
+
+    get '/' do
+      erb :index
+    end
+
+    def logged_in?
+      !!session[:user_id]
+    end
+
+    get '/signup' do
+      if self.logged_in?
+          redirect "/tweets"
+      else
+        erb :'/users/create_user'
+      end
+    end
+
+    post '/signup' do
+      if params[:username]=="" || params[:email]==""|| params[:password]==""
+        redirect "/signup"
+      else
+        @user = User.create(params)
+        session[:user_id] = @user.id
+        redirect "/tweets"
+      end
+    end
+
+    get '/login' do
+        if self.logged_in?
+          redirect "/tweets"
+        else
+          erb :'/users/login'
+        end
+    end
+
+    post '/login' do
+      @user = User.find_by(username: params[:username])
+      if @user && @user.authenticate(params[:password])
+        session[:user_id] = @user.id
+        redirect "/tweets"
+      else
+        redirect "/login"
+      end
+
+    end
+
+    get '/logout' do
+      if self.logged_in?
+        session.clear
+        redirect "/login"
+      else
+        redirect "/"
+      end
+    end
+
+    get '/tweets' do
+
+    if logged_in?
+      @tweets = Tweet.all
+      @user = User.find(session[:user_id])
+      erb :'/tweets/tweets'
+    else
+      redirect "/login"
+    end
   end
 
-end
\ No newline at end of file
+    get '/users/:slug' do
+      @user = User.find_by_slug(params[:slug])
+      erb :'/users/show'
+    end
+
+    get '/tweets/new' do
+      if self.logged_in?
+        erb :'/tweets/create_tweet'
+      else
+        redirect "/login"
+      end
+      end
+
+    post '/tweets/new' do
```

Comment:
Typically, following [RESTful convention](https://learn.co/tracks/full-stack-web-development-v5/sinatra/activerecord/sinatra-restful-routes), tweet creation would be posted to `/tweets` instead of `/tweets/new`.